On Windows, fix request_redraw() related panics

filnet opened the pull request:
~~Simplifies the message loops and event handling considerably~~
~~Strongly inspired from the X11 platform~~

~~Partially reverts https://github.com/rust-windowing/winit/commit/6a330a2894873d29fbbfdeebfc1a215577213996~~

 These panics were introduced by https://github.com/rust-windowing/winit/commit/6a330a2894873d29fbbfdeebfc1a215577213996

Fixes https://github.com/rust-windowing/winit/issues/1391
Fixes https://github.com/rust-windowing/winit/issues/1400
Fixes https://github.com/rust-windowing/winit/issues/1466
Fixes #1469
Probably fixes other related issues

The commit mentioned above introduced a few critical regressions.
I did quite a bit of forensic (looking at 0.19.1, 0.20.0-alpha1) to understand what was going on around the handling of request_redraw(). I also looked at how other platforms handle it.

~~I tried really hard to fix the issues using the current approach of posting WM_PAINT messages but got nowhere. Too many corner cases.
In the end I decided to go for the X11 approach of using a HashSet to keep track of windows requiring a redraw and things got much simpler.~~

This PR does not fundamentally change how request_redraw() is handled.
It still calls `RedrawWindow` which sends a `WM_PAINT` message. Which means that:
- multiple `WM_PAINT` messages are coalesced (for example when calling request_redraw while resizing or calling request_redraw multiple times),
- `RedrawRequested` events are always sent when handling a `WM_PAINT` message.

I tried to be very thorough in testing:
- tested all control flow modes (Poll, Wait and WaitUntill) in both the non-modal and modal loops
- made sure that redraw requests are properly coalesced (in particular when resizing)
- tested multi window and fullscreen

There are a few untested corner cases. See https://github.com/rust-windowing/winit/issues/1429 for more details.

And always kept the specs in a corner of my mind:
- https://docs.rs/winit/0.21.0/winit/window/struct.Window.html#method.request_redraw
- https://docs.rs/winit/0.21.0/winit/event/index.html

- [x] Tested on all platforms changed
- [x] Compilation warnings were addressed
- [x] `cargo fmt` has been run on this branch
- [ ] `cargo doc` builds successfully
- [x] Added an entry to `CHANGELOG.md` if knowledge of this change could be valuable to users
- [ ] Updated documentation to reflect any user-facing changes, including notes of platform-specific behavior
- [ ] Created or updated an example program if it would help users understand this functionality
- [ ] Updated [feature matrix](https://github.com/rust-windowing/winit/blob/master/FEATURES.md), if new features were added or implemented

filnet commented:
I think I am done for now. All reported panics are not reproducible anymore.

Imberflur commented:
I'm guessing it probably also fixes #1469

Imberflur commented:
Windows user experienced some panics.
```
PanicInfo: panicked at 'buffering RedrawRequested event', <::std::macros::panic macros>:2:4
.
.
.
  28: std::panicking::begin_panic_fmt
             at /rustc/58b834344fc7b9185e7a50db1ff24e5eb07dae5e\/src\libstd\panicking.rs:332
  29: winit::platform_impl::platform::event_loop::runner::EventLoopRunner<T>::new_events
  30: winit::platform_impl::platform::event_loop::runner::ELRShared<T>::send_event
  31: winit::platform_impl::platform::event_loop::public_window_callback
  32: DefSubclassProc
  33: DefSubclassProc
  34: CallWindowProcW
  35: CallWindowProcW
  36: glPushClientAttrib
  37: CallWindowProcW
  38: DispatchMessageW
  39: IsWindowVisible
  40: KiUserCallbackDispatcher
  41: NtUserPeekMessage
  42: PeekMessageW
  43: PeekMessageW
  44: winit::platform_impl::platform::event_loop::EventLoop<T>::run_return
  45: winit::platform_impl::platform::event_loop::EventLoop<T>::run
  46: winit::event_loop::EventLoop<T>::run
```
and when switching to a different window:
```rs
PanicInfo: panicked at 'non-redraw event in redraw phase: Some(WindowEvent { window_id: WindowId(WindowId(0x30594)), event: Focused(false) })', <::std::macros::panic macros>:5:6
.
.
.
   5: std::panicking::begin_panic_fmt
             at /rustc/58b834344fc7b9185e7a50db1ff24e5eb07dae5e\/src\libstd\panicking.rs:332
   6: winit::platform_impl::platform::event_loop::runner::EventLoopRunner<T>::new_events
   7: winit::platform_impl::platform::event_loop::runner::ELRShared<T>::send_event
   8: winit::platform_impl::platform::event_loop::public_window_callback
   9: DefSubclassProc
  10: DefSubclassProc
  11: CallWindowProcW
  12: CallWindowProcW
  13: glPushClientAttrib
  14: CallWindowProcW
  15: DispatchMessageW
  16: IsWindowVisible
  17: KiUserCallbackDispatcher
  18: NtUserPeekMessage
  19: PeekMessageW
  20: PeekMessageW
  21: winit::platform_impl::platform::event_loop::EventLoop<T>::run_return
  22: winit::platform_impl::platform::event_loop::EventLoop<T>::run
  23: winit::event_loop::EventLoop<T>::run
```
and
```
PanicInfo: panicked at 'non-redraw event in redraw phase: Some(DeviceEvent { device_id: DeviceId(DeviceId(65590)), event: Motion { axis: 1, value: -1.0 } })', ........\src\platform_impl\windows\event_loop\runner.rs:364:17

Backtrace:
   0: backtrace::backtrace::trace
   1: backtrace::capture::Backtrace::new
   2: cpal::traits::HostTrait::output_devices::supports_output
   3: std::panicking::rust_panic_with_hook
             at /rustc/58b834344fc7b9185e7a50db1ff24e5eb07dae5e\/src\libstd\panicking.rs:474
   4: std::panicking::begin_panic_handler
             at /rustc/58b834344fc7b9185e7a50db1ff24e5eb07dae5e\/src\libstd\panicking.rs:378
   5: std::panicking::begin_panic_fmt
             at /rustc/58b834344fc7b9185e7a50db1ff24e5eb07dae5e\/src\libstd\panicking.rs:332
   6: winit::platform_impl::platform::event_loop::runner::EventLoopRunner<T>::new_events
   7: winit::platform_impl::platform::event_loop::runner::ELRShared<T>::send_event
   8: winit::platform_impl::platform::event_loop::normalize_pointer_pressure
   9: DefSubclassProc
  10: DefSubclassProc
  11: CallWindowProcW
  12: DispatchMessageW
  13: IsWindowVisible
  14: KiUserCallbackDispatcher
  15: NtUserMessageCall
  16: GetWindowTextW
  17: MapWindowPoints
  18: IsCompositionActive
  19: Ordinal96
  20: IsCompositionActive
  21: IsCompositionActive
  22: GetWindowTextW
  23: DefSubclassProc
  24: DefSubclassProc
  25: CallWindowProcW
  26: CallWindowProcW
  27: glPushClientAttrib
  28: CallWindowProcW
  29: DispatchMessageW
  30: IsWindowVisible
  31: KiUserCallbackDispatcher
  32: NtUserMessageCall
  33: GetWindowTextW
  34: MapWindowPoints
  35: IsCompositionActive
  36: Ordinal96
  37: IsCompositionActive
  38: IsCompositionActive
  39: GetWindowTextW
  40: CallWindowProcW
  41: CallWindowProcW
  42: DefSubclassProc
  43: DefSubclassProc
  44: DefSubclassProc
  45: DefSubclassProc
  46: CallWindowProcW
  47: CallWindowProcW
  48: glPushClientAttrib
  49: CallWindowProcW
  50: DispatchMessageW
  51: winit::platform_impl::platform::event_loop::EventLoop<T>::run_return
  52: winit::platform_impl::platform::event_loop::EventLoop<T>::run
  53: winit::event_loop::EventLoop<T>::run
```

filnet commented:
@Imberflur what was the user doing and which Windows version was used ?

I think I understand the issue but can't reproduce it. See https://github.com/rust-windowing/winit/issues/1429#issuecomment-591868164.

And where can I find the source code to take a look ?

Imberflur commented:
Here is the event loop code: https://gitlab.com/veloren/veloren/-/blob/zesterer/winit-20/voxygen/src/run.rs
Maybe it is easy to trigger because it lingers in the `MainEventsCleared` handler?
@filnet

filnet commented:
I was able to get a panic by sleeping when handling `MainEventsClear`.

@Imberflur your first panic has this strange line:
>  12: winit::platform_impl::platform::event_loop::public_window_callback<()>
             at src\platform_impl\windows\event_loop.rs:0

Line 0 ???

filnet commented:
@Imberflur not sure it is going to address your issues but in your game loop, you should:
1/ call window.request_redraw() from `MainEventsCleared` if a redraw is needed
2/ move your rendering code to `RedrawRequested`.
This is the 'correct' way to write a winit event loop and will change the place/time where you sleep (might workaround the issue).

Imberflur commented:
@filnet I avoided using the "correct" method for redrawing because it kept the code simpler/closer to the original structure and I don't have the need currently to respond to external redraw requests. Nevertheless, I doubt it would circumvent the issue since time would still be spent in `MainEventsCleared` doing update logic.

filnet commented:
I just pushed a small fix that should address some of the panics.

aloucks commented on src/platform_impl/windows/event_loop.rs:
```diff
@@ -638,13 +632,6 @@ unsafe extern "system" fn public_window_callback<T: 'static>(
         }
 
         winuser::WM_NCLBUTTONDOWN => {
-            // jumpstart the modal loop
```
This `RedrawWindow` and the `PostMessageW` below it were to prevent the main loop from pausing when the user clicks the title bar.

filnet commented:
I am more or less done now. But there are two aspects that I am not too happy about.

1/ I added a few panic calls to catch some undefined use cases.
Those involve calling window functions that change the window state (like `set_inner_size`) when handling `MainEventsCleared`, `RedrawRequested` or `RedrawEventsCleared`. Are those allowed and, if yes, how should they behave ?

2/ There is a handful of calls to `winuser::SetWindowPos` that were all immediately followed by calls to `winuser::RedrawWindow`. I replaced all `winuser::RedrawWindow` calls by `winuser::InvalidateRgn`.
I think the calls to `winuser::InvalidateRgn` can be removed.
Anyone knows why those `winuser::RedrawWindow` calls were added in the first place. Reading the `winuser::SetWindowPos` documentation leads me to beleive that this methods triggers redraw when appropriate.

aloucks commented:
> I added a few panic calls to catch some undefined use cases. Those involve calling window functions that change the window state (like set_inner_size) when handling `MainEventsCleared`, `RedrawRequested` or `RedrawEventsCleared`. Are those allowed and, if yes, how should they behave ?

I believe the desired usage is that you enter the eventloop and it takes over as your `main`. There's an [event loop extension trait](https://docs.rs/winit/0.21.0/winit/platform/desktop/trait.EventLoopExtDesktop.html#tymethod.run_return) that let's you break out, but I don't think this is the preferred way.

With that said, I'd imagine that the expectation for state change operations like `set_inner_size` is that they should not panic and should be honored.

filnet commented:
> With that said, I'd imagine that the expectation for state change operations like `set_inner_size` is that they should not panic and should be honored.

Yep and that's the whole point: because of how winit does things on Windows handling set_inner_size calls (and other similar winit::Windows functions) that originate from user code  while handling MainEventsCleared is broken and hard to fix.

There is a sort of catch 22 situation. Calling set_inner_size will _immediately_ emit a WM_SIZE and WM_PAINT message. The WM_SIZE message will generate Event::Resized event that will get buffered. It is buffered because you can't dispacth an event while another one is being dispatched (in our case we are currently dispatching MainEventsCleared). On the other hand, for WM_PAINT, you can't buffer the RedrawRequested event because apparently, on Windows, you should (must?) redraw while handling the WM_MESSAGE.
A fix would be to defer the "whole" set_inner_size() call to after dispatching the RedrawEventsCleared (i.e. when the current loop iteration is finished) but I don't know how to do that.

Anyways, if I change the panic() call that is done when detecting this situation to warn() call then things sort of work. The call is honored (i.e. the window is resized) but the user sees strange sequence of events.

filnet commented:
To summarize the situation:

winit forwards native UI events and also injects a few synthetic events (`NewEvents`, `MainEventsCleared`, `RedrawEventsCleared`).

The synthetic events act as some sort of checkpoints.
For instance `MainEventsCleared` tells the user code to that all current events have been dispatched and that it should get ready to redraw. If the user code calls code that triggers new events then we are in a bad situation. Like wise, changing the window state while rendering in not a good thing.

Unfortunately, winit cannot forbid user code from changing the window state from `MainEventsCleared` or `RedrawRequested`.  The question then is should we fail badly when it happens or should just warn and break softly.

The doc should be updated to mention that some use cases are discouraged and have undefined behaviors.

Osspial commented:
>  A fix would be to defer the "whole" `set_inner_size` call to after dispatching the `RedrawEventsCleared` (i.e. when the current loop iteration is finished) but I don't know how to do that. Edit: that fix is questionable...

We could change `set_inner_size` to post a custom message to the event queue, then perform the resize logic in that event handler. We'd have to update the documentation to reflect that, though.

filnet commented:
> We could change `set_inner_size` to post a custom message to the event queue, then perform the resize logic in that event handler. We'd have to update the documentation to reflect that, though.
> 
> EDIT: the `UpdateWindow->InvalidateRgn` change doesn't fix that issue?

`set_inner_size` is just one of the many state changing function exposed by `Window` and no the `UpdateWindow->InvalidateRgn` does not fix the issue.

There is no proper fix in my opinion. The [doc](https://docs.rs/winit/0.21.0/winit/event/enum.Event.html) tells you what you're supposed to do when handling `MainEventsCleared`.

```
MainEventsCleared
[−]

Emitted when all of the event loop's input events have been processed and redraw processing is about to begin.

This event is useful as a place to put your code that should be run after all state-changing events have been handled and you want to do stuff (updating state, performing calculations, etc) that happens as the "main body" of your event loop. If your program draws graphics, it's usually better to do it in response to Event::RedrawRequested, which gets emitted immediately after this event.
```
To me, it is implied that doing anything else will result in undefined behavior.

The problem is that the user code has full access to the window. Ideally he it should have access to a mutable window when handling regular events and to a non mutable events when handling `MainEventsCleared` and `RedrawRequested`. But that is currently not possible.

An extreme example: In `RedrawRequested` a user could change the window size in the middle of rendering a complex scene. No one would do that because it is crazy but, yet, it is currently possible to do that.

This quite long video mentions this issue (in another but similar context) : https://www.youtube.com/watch?v=4YTfxresvS8

filnet commented:
I think what is allowed to be done in `MainEventsCleared` is not specific to Windows. We should discuss it with the maintainers of the other platforms.

Also note that all winit's example call state changing function (set_fullscreen, etc) when handling keyboard events. They never do it from `MainEventsCleared`. 
And a user would expect the set_inner_size() call done from `MainEventsCleared` to be effective immediately in the following `RedrawRequested`. But how do we send the `Resized` event. We just can't because we are in `MainEventsCleared` state...

Now the real question, since we can't prevent it, is how to handle it : crash, warn or fumble.
And, again, how do the other platforms handle it ?

Osspial commented on src/platform_impl/windows/event_loop.rs:
```diff
@@ -223,67 +223,60 @@ impl<T: 'static> EventLoop<T> {
+
+                    if msg.message == winuser::WM_PAINT {
+                        break;
+                    }
```
Now that this `WM_PAINT` check moved after `TranslateMessage` and `DispatchMessage`, the first `WM_PAINT` event will be handled before `main_events_cleared` is called. This won't cause any bugs because the runner automatically dispatches `MainEventsCleared` before `RedrawRequested`, but this code still looks like it should be buggy upon first glance. Could you add a comment explaining why it works?

filnet replied:
done

Osspial commented on src/platform_impl/windows/event_loop/runner.rs:
```diff
@@ -496,57 +438,14 @@ impl<T> EventLoopRunner<T> {
-                self.runner_state = RunnerState::Idle(wait_start)
-            }
+            // If we *weren't* handling redraws, we don't have to do anything.
+            _ => (),
```
I'm not entirely comfortable with just doing nothing on this branch - if it's invoked, there's probably a bug higher up in the codebase caused `redraw_events_cleared` to be called at the wrong time. Not sure if we should panic or warn about this, but whichever way we go it shouldn't be silent.

filnet replied:
I think this handles the case were a Windows message is sent but we don't handle it at all (i.e. it does not translate to a winit event). There's a bunch of those. When one those arrives we will be woken up but we must do nothing.

Osspial replied:
That's the case for a lot of the `RunnerState`s, but not all of them. If it's called during the `HandlingEvents` phase, this shouldn't be a no-op.

filnet replied:
PS: we can arrive here in `DeferredNewEvents` state which is ok and expected.
I can make the core more explicit by letting only `DeferredNewEvents` pass through and panic on other states. Or add a comment. Let me know.

filnet replied:
made redraw_events_cleared more explicit.

Osspial replied:
Looks good, thanks!

Osspial commented:
@filnet My stance is that calling any window state-changing function during `MainEventsCleared` should be allowable behavior - if the documentation doesn't say that a particular usage pattern is explicitly wrong, we should do our best to handle it gracefully. I think it's perfectly reasonable for us to add a clause to the size-changing functions that specifies that the size change doesn't take effect until the `Resized` event is received.

filnet commented:
@Osspial I'll try to put this PR in perspective.

I came to winit as I decided to learn Rust and port a pet project (a Voxel engine in Monogame) to it.
I quickly found winit and started to use it to write a game loop. I started with version 0.20.0 and then 0.21.0.

I then fell into the rabbit hole of trying to fix the issues introduced in this major release.
I tried a few "new" approaches and went full circle to in the end just tighten your code.
I believe your current architecture can be made to work and, I hope, this PR proves it.
I might have over tightened the code in a few places (i.e. made more obscure) ;)

AFAIK, the two major issues remaining, in 0.21.0 and this PR, are:
1/ changing window state in `MainEventsCleared`, etc... as discussed.
2/ the event loop can be frozen from the title bar (right clicking the title bar, clicking the window icon, ...)

Osspial commented:
Alright, this looks good. Thanks for the work on this! I'll go ahead and merge this now, since it seems to be a bit more stable than #1496 is at present.

That being said, I still intend to merge #1496 eventually, once the issues with it have been addressed. There are a good number of concrete technical reasons to go to the additional lengths that PR does, but it'll be better to explain those in that particular thread.
